Entities/TerrainDebris.cpp

```diff
@@ -84,104 +103,147 @@ namespace RTE {
 	}
 
 /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
-	void TerrainDebris::ApplyDebris(SLTerrain *terrain) {
-		RTEAssert(!m_Bitmaps.empty() && m_BitmapCount > 0, "No bitmaps loaded for terrain debris!");
-
-		BITMAP *terrBitmap = terrain->GetFGColorBitmap();
+	bool TerrainDebris::GetPiecePlacementPosition(SLTerrain *terrain, Box &positionCheckBox) const {
 		BITMAP *matBitmap = terrain->GetMaterialBitmap();
-
-		// How many pieces of debris we're spreading out.
-		unsigned int terrainWidth = terrBitmap->w;
-		unsigned int pieceCount = static_cast<int>((static_cast<float>(terrainWidth) * c_MPP) * m_Density);
-
-		// First is index in the bitmap array, Second is blit location
-		std::list<std::pair<int, Vector>> piecesToPlace;
-		Vector location;
-		Box pieceBox;
-
-		unsigned char checkPixel;
-
-		acquire_bitmap(terrBitmap);
-		acquire_bitmap(matBitmap);
-
-		for (unsigned int piece = 0; piece < pieceCount; ++piece) {
-			bool place = false;
-			unsigned int currentBitmap = RandomNum<int>(0, m_BitmapCount - 1);
-			RTEAssert(currentBitmap >= 0 && currentBitmap < m_BitmapCount, "Bitmap index is out of bounds!");
-
-			pieceBox.SetWidth(static_cast<float>(m_Bitmaps.at(currentBitmap)->w));
-			pieceBox.SetHeight(static_cast<float>(m_Bitmaps.at(currentBitmap)->h));
-
-			int x = RandomNum<int>(0, terrainWidth);
-			int y = 0;
-			int depth = RandomNum(m_MinDepth, m_MaxDepth);
-
-			while (y < terrBitmap->h) {
-				// Find the air-terrain boundary
-				for (; y < terrBitmap->h; ++y) {
-					checkPixel = _getpixel(matBitmap, x, y);
-					// Check for terrain hit
-					if (checkPixel != g_MaterialAir) {
-						if (checkPixel == m_TargetMaterial.GetIndex()) {
-							place = true;
-							break;
-						// If we didn't hit target material, but are specified to, then don't place
-						} else if (m_OnlyOnSurface) {
-							place = false;
-							break;
+		int posX = RandomNum(0, matBitmap->w);
+		int depth = RandomNum(m_MinDepth, m_MaxDepth);
+		int buriedDepthOffset = m_OnlyBuried ? static_cast<int>(positionCheckBox.GetHeight() * 0.6F) : 0;
+		int prevMaterialCheckPixel = -1;
+
+		bool scanForOverhang = m_DebrisPlacementMode == DebrisPlacementMode::OnOverhangOnly || m_DebrisPlacementMode == DebrisPlacementMode::OnCavityOverhangOnly || m_DebrisPlacementMode == DebrisPlacementMode::OnOverhangAndCavityOverhang;
+
+		// For overhangs scan from the bottom so it's easier to detect.
+		for (int surfacePosY = 0, overhangPosY = matBitmap->h - 1; surfacePosY < matBitmap->h && overhangPosY > 0; surfacePosY++, overhangPosY--) {
+			int posY = scanForOverhang ? overhangPosY : surfacePosY;
+			int materialCheckPixel = _getpixel(matBitmap, posX, posY);
+			if (materialCheckPixel != MaterialColorKeys::g_MaterialAir) {
+				// TODO: Adding depth here will properly place debris only on target material, rather than any random pixel within depth range from the original target material pixel.
```

Comment:
Man this is so long to write out y u do dis? I know what this means and i'll deal with it one day.
This bug (imo) exists from the dawn of time.

Basically, when target material pixel is encountered random depth is added to offset the debris piece.
This means that if target material is `Dirt` and the pixel is at `Y = 100`, when `depth = whatever` is added the actual pixel the piece will be placed at might not be of target material because no extra checking is performed.
If you uncomment the extra check, it will work as it is supposed to (imo), only placing debris on the target material.

Multi target materials is needed to sorta preserve the current behavior where debris like boulders are targeted to e.g `Dirt` but are also placed on `Rock` that is usually deeper because they are offset by depth, but at the same time preventing placing them to materials that aren't targeted (e.g you have a thin layer of ice between your dirt and rock, the current behavior will place boulders on ice, multi target will only place on dirt and rock).

Reply:
Ahhhhh thank you for a genuinely great, clear explanation. This also explains some debris generation visual weirdness I've seen in game at times.
Cool, feel free to resolve.